File: backend/src/controllers/authController.js
import User from '../models/User.js';
import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import Session from '../models/Session.js';

export const login = async (req, res) => {
  console.log('[AUTH][LOGIN] Tentative de connexion pour', req.body.username);
  
  try {
    const { username, password } = req.body;
    if (!username || !password) {
      console.log('[AUTH][LOGIN] Champs manquants');
      return res.status(400).json({ message: 'Nom d\'utilisateur et mot de passe requis.' });
    }

    const user = await User.findOne({ username, isActive: true })
      .populate('assignedStores', 'name isActive'); // NOUVEAU

    if (!user) {
      console.log(`[AUTH][LOGIN] Utilisateur non trouvé : ${username}`);
      return res.status(401).json({ message: 'Identifiants invalides.' });
    }

    if (!user.isActive) {
      console.log(`[AUTH][LOGIN] Utilisateur inactif : ${username}`);
      return res.status(403).json({ message: 'Compte désactivé.' });
    }

    const isMatch = await bcrypt.compare(password, user.passwordHash);
    if (!isMatch) {
      console.log(`[AUTH][LOGIN] Mot de passe incorrect pour : ${username}`);
      return res.status(401).json({ message: 'Identifiants invalides.' });
    }

    // Vérifier que l'utilisateur a au moins un magasin assigné (sauf pour super-admin)
    if (user.role !== 'super-admin' && (!user.assignedStores || user.assignedStores.length === 0)) {
      console.log('[AUTH][LOGIN] Échec - aucun magasin assigné');
      return res.status(403).json({ 
        message: 'Aucun magasin assigné à cet utilisateur.' 
      });
    }

    // Générer le JWT avec les magasins assignés
    const expiresIn = '8h';
    const token = jwt.sign(
      { 
        userId: user._id, 
        username: user.username, 
        role: user.role,
        assignedStores: user.assignedStores ? user.assignedStores.map(s => s._id) : [] // Gestion du cas undefined
      },
      process.env.JWT_SECRET,
      { expiresIn }
    );

    // Création de la session
    const userAgent = req.headers['user-agent'] || '';
    const ip = req.headers['x-forwarded-for'] || req.connection.remoteAddress || '';
    const decoded = jwt.decode(token);
    const expiresAt = new Date(decoded.exp * 1000);
    
    await Session.create({
      userId: user._id,
      token,
      deviceInfo: {
        userAgent,
        ip,
        platform: req.body.platform || ''
      },
      expiresAt
    });

    console.log(`[AUTH][LOGIN] Connexion réussie pour : ${username} (${user.role})`);
    
    res.json({ 
      token, 
      user: { 
        id: user._id,
        username: user.username, 
        role: user.role,
        assignedStores: user.assignedStores || [] // Gestion du cas undefined
      } 
    });
  } catch (error) {
    console.error('[AUTH][LOGIN] Erreur serveur :', error.message);
    res.status(500).json({ message: 'Erreur serveur.', error: error.message });
  }
};

export const logout = async (req, res) => {
  try {
    const authHeader = req.headers['authorization'];
    const token = authHeader && authHeader.split(' ')[1];
    if (!token) {
      return res.status(401).json({ message: 'Token manquant.' });
    }
    // Désactiver la session
    const session = await Session.findOne({ token, isActive: true });
    if (!session) {
      return res.status(400).json({ message: 'Session non trouvée ou déjà déconnectée.' });
    }
    session.isActive = false;
    await session.save();
    return res.json({ message: 'Déconnexion réussie.' });
  } catch (error) {
    console.error('[AUTH][LOGOUT] Erreur serveur :', error.message);
    res.status(500).json({ message: 'Erreur serveur.', error: error.message });
  }
};

export const verifySession = async (req, res) => {
  try {
    const authHeader = req.headers['authorization'];
    const token = authHeader && authHeader.split(' ')[1];
    if (!token) {
      return res.status(401).json({ valid: false, message: 'Token manquant.' });
    }
    
    // Vérification cryptographique du token
    const user = jwt.verify(token, process.env.JWT_SECRET);
    
    // Vérification de la session active en base
    const session = await Session.findOne({ 
      token, 
      isActive: true,
      expiresAt: { $gt: new Date() }
    });
    
    if (!session) {
      return res.status(401).json({ valid: false, message: 'Session inactive ou expirée.' });
    }
    
    return res.json({ 
      valid: true, 
      user: { username: user.username, role: user.role },
      sessionInfo: {
        lastActivity: session.lastActivity,
        expiresAt: session.expiresAt,
        deviceInfo: session.deviceInfo
      }
    });
  } catch (error) {
    if (error.name === 'TokenExpiredError') {
      return res.status(401).json({ valid: false, message: 'Token expiré.' });
    } else if (error.name === 'JsonWebTokenError') {
      return res.status(403).json({ valid: false, message: 'Token invalide.' });
    } else {
      console.error('[AUTH][VERIFY] Erreur serveur :', error.message);
      return res.status(500).json({ valid: false, message: 'Erreur serveur.' });
    }
  }
};

export const getSessionInfo = async (req, res) => {
  try {
    // req.session est disponible grâce au middleware verifyToken
    const session = req.session;
    const user = req.user;
    
    return res.json({
      user: { username: user.username, role: user.role },
      session: {
        lastActivity: session.lastActivity,
        expiresAt: session.expiresAt,
        deviceInfo: session.deviceInfo,
        createdAt: session.createdAt
      }
    });
  } catch (error) {
    console.error('[AUTH][SESSION_INFO] Erreur serveur :', error.message);
    res.status(500).json({ message: 'Erreur serveur.' });
  }
}; 

File: frontend/lib/controllers/auth_controller.dart
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../services/auth_service.dart';
import '../services/store_service.dart'; // NOUVEAU
// NOUVEAU
// NOUVEAU
import 'dart:convert';

class AuthController {
  final AuthService _authService = AuthService();
  final StoreService _storeService = StoreService(); // NOUVEAU

  Future<Map<String, dynamic>> login(String username, String password) async {
    debugPrint('[CONTROLLER][AuthController] Appel login pour $username');
    final result = await _authService.login(username, password);
    final statusCode = result['statusCode'];
    final body = result['body'];
    debugPrint(
      '[CONTROLLER][AuthController] Status: ${statusCode}, Body: ${body}',
    );
    if (statusCode == 200) {
      final data = jsonDecode(body);
      final token = data['token'];
      final userData = data['user'];
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString('jwt_token', token);
      await prefs.setString('user_role', userData['role']);
      await prefs.setString('username', userData['username']);
      // Ajout stockage ID utilisateur
      await prefs.setString('user_id', userData['id'] ?? userData['_id']);

      // Stocker les magasins assignés
      if (userData['assignedStores'] != null) {
        final storesJson = jsonEncode(userData['assignedStores']);
        await prefs.setString('assigned_stores', storesJson);
        debugPrint(
          '[CONTROLLER][AuthController] Magasins assignés stockés: ${userData['assignedStores'].length}',
        );
      }

      debugPrint(
        '[CONTROLLER][AuthController] Token et rôle stockés pour ${userData['username']} (${userData['role']})',
      );
      return {
        'success': true,
        'role': userData['role'],
        'username': userData['username'],
        'assignedStores': userData['assignedStores'] ?? [], // NOUVEAU
      };
    } else {
      final data = jsonDecode(body);
      debugPrint(
        '[CONTROLLER][AuthController] Échec login: ${data['message']}',
      );
      return {
        'success': false,
        'message': data['message'] ?? 'Identifiants invalides',
      };
    }
  }

  Future<Map<String, dynamic>> logout() async {
    debugPrint('[CONTROLLER][AuthController] Appel logout');
    try {
      final prefs = await SharedPreferences.getInstance();
      final token = prefs.getString('jwt_token');

      if (token != null) {
        // Appel au serveur pour invalider la session
        final result = await _authService.logout(token);
        debugPrint(
          '[CONTROLLER][AuthController] Logout serveur: ${result['statusCode']}',
        );
      }

      // Nettoyage des données locales
      await prefs.clear();
      debugPrint('[CONTROLLER][AuthController] Données locales nettoyées');

      return {'success': true, 'message': 'Déconnexion réussie'};
    } catch (e) {
      debugPrint('[CONTROLLER][AuthController] Erreur logout: ${e.toString()}');
      // Même en cas d'erreur, on nettoie les données locales
      final prefs = await SharedPreferences.getInstance();
      await prefs.clear();
      return {'success': true, 'message': 'Déconnexion locale effectuée'};
    }
  }

  Future<Map<String, dynamic>> verifySession() async {
    debugPrint('[CONTROLLER][AuthController] Vérification de session');
    try {
      final prefs = await SharedPreferences.getInstance();
      final token = prefs.getString('jwt_token');

      if (token == null) {
        return {'valid': false, 'message': 'Aucun token trouvé'};
      }

      final result = await _authService.verifySession(token);
      final statusCode = result['statusCode'];
      final body = result['body'];

      if (statusCode == 200) {
        final data = jsonDecode(body);
        return {
          'valid': true,
          'user': data['user'],
          'sessionInfo': data['sessionInfo'],
        };
      } else {
        final data = jsonDecode(body);
        return {
          'valid': false,
          'message': data['message'] ?? 'Session invalide',
        };
      }
    } catch (e) {
      debugPrint(
        '[CONTROLLER][AuthController] Erreur vérification: ${e.toString()}',
      );
      return {'valid': false, 'message': 'Erreur de vérification'};
    }
  }
}
